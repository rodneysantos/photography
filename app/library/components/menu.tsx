interface MenuProps {
	// The class name to apply to the div element.
	className?: string;
}

export const Menu = ({ className }: MenuProps) => {
	return (
		<div className={className}>
			<div className="flex flex-row justify-center">
				<div className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900">
					black and white
				</div>

				<div className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900">
					color
				</div>

				<div className="px-2 py-1 mx-2 text-center border-2 cursor-pointer bg-neutral-900 text-neutral-100 border-neutral-900">
					low-key
				</div>
			</div>
		</div>
	);
};
