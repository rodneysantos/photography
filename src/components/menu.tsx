import { createSignal } from "solid-js";
import { JSX } from "solid-js/jsx-runtime";

interface MenuItem {
	// The id of the item.
	id: number;

	// The name of the item.
	name: Item;
}

const menuItems: MenuItem[] = [
	{ id: 0, name: "black and white" },
	{ id: 1, name: "color" },
	{ id: 2, name: "low-key" },
];

export type Item = "black and white" | "color" | "low-key";

export interface MenuProps extends JSX.HTMLAttributes<HTMLDivElement> {
	// The ids of the active items.
	defaultActiveItems: Item[];

	// The callback to call when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;

	// Whether or not to show the menu.
	show: boolean;
}

export const Menu = (props: MenuProps) => {
	const [activeItems, setActiveItems] = createSignal(props.defaultActiveItems);

	// Toggle the active state of an item.
	// If the item is active, remove it from the active items.
	// If the item is not active, add it to the active items.
	function handleToggleItem(name: Item) {
		const updatedActiveItems = activeItems().includes(name)
			? activeItems().filter((item) => item !== name)
			: [...activeItems(), name];

		setActiveItems(updatedActiveItems);
		props.onActiveItemsChange(updatedActiveItems);
	}

	return (
		<div
			data-testid="menu"
			class="transition-opacity will-change-auto"
			classList={{ "opacity-0": !props.show }}
		>
			<div class="flex flex-row justify-center">
				{menuItems.map((item) => {
					const isActive = activeItems().includes(item.name);

					return (
						<div
							data-testid={item.name}
							onClick={() => handleToggleItem(item.name)}
							class="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900"
							classList={{ "bg-neutral-900 text-neutral-100": isActive }}
						>
							{item.name}
						</div>
					);
				})}
			</div>
		</div>
	);
};
