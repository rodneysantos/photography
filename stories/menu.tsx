import { Menu } from "../app/library/components/menu";

export const MenuSB = () => {
	return <Menu />;
};
