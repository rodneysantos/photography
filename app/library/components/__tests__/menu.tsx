import renderer from "react-test-renderer";
import { Menu } from "../menu";

describe("menu", () => {
	it("renders correctly", () => {
		// arrange
		const tree = renderer.create(<Menu />).toJSON();

		// assert
		expect(tree).toMatchSnapshot();
	});
});
