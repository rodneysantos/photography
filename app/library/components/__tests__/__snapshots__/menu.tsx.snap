// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu renders correctly 1`] = `
<div>
  <div
    className="flex flex-row justify-center"
  >
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900"
    >
      black and white
    </div>
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900"
    >
      color
    </div>
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer bg-neutral-900 text-neutral-100 border-neutral-900"
    >
      low-key
    </div>
  </div>
</div>
`;
